### DISCUSSION TITLE
Any reason for jax.lax.cond not playing nice with jacobian

### DISCUSSION THREAD

**Discussion by samskiter** [jax-ml/jax#17341] [General] [answered]
I'm currently trying to reproduce a simpler version of a bug I'm facing in my code - essentially a jax.lax.cond is killing my gradients in my jacobian. Before I go mad, are there any gotchas I might be missing?

I have a class that implements a method something like this:

```
def get_error(self) -> float:

    return jax.lax.cond(
        self.some_class_value > 1.0,
        self._calculate_error,
        self._calculate_other_error,
    )
```

This is used in some higher error function, which I then take a `jacfwd` of.

I noticed I was seeing zero gradients in parts of my jacobian in places where I saw them on an old reference implementation. I also noticed that if I removed the conditional, those gradients came back and matched the reference implementation.

The following worked (In my debugging case I knew the conditional evaluated false):

```
def get_error(self) -> float:
    self._calculate_other_error()
```

The following did not work:

```
def get_error(self) -> float:

    return jax.lax.cond(
        False,
        self._calculate_error,
        self._calculate_other_error,
    )
```

Nor did the following:

```
def get_error(self) -> float:

    return jax.lax.cond(
        True,
        self._calculate_other_error,
        self._calculate_other_error,
    )
```

It seems that merely the presence of the `cond` is throwing something off in the jacobian

As I mentioned, I'm still trying to reproduce this outside of my code base (where things seem to be behaving a little better), but before I go mad, are there any gotchas I ought to be looking out for ?

Thanks!

**Comment by jakevdp** [comment]
What do the false branches return? This might be some manifestation of this issue [JAX FAQ: gradients contain NaN when using `where`](https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where). The gradient of a `cond` is basically computed as `1 * grad(true_branch)(x) + 0 * grad(false_branch(x))`, so if `false_branch` returns something like `NaN` or `inf`, it can cause issues with the overall gradient.

In any case, if you're able to put together a minimal reproduction, it would be easier to help.

**Reply by samskiter** [reply]
Thanks for the reply - still trying to get this minimal repro together - I read the FAQ article you linked to but thats for reverse mode differentiation - I'm using jacfwd (infact I've tried with both)

Anyway, the branch just returns a number: 0.09181296626905545 to be precise, so it's not NaN or inf.

**Reply by jakevdp** [reply]
Another guess: does your class do any in-place mutation (i.e. methods that have side-effects of the form `self.attr = ...`). If your functions are impure, I could imagine that causing this kind of issue in autodiff.

**Comment by samskiter** [comment]
Definitely have lots of impure stuff but I've registered pytrees.
Flatten/unflatten not being called though :/

**Reply by jakevdp** [reply]
Registering pytrees is not enough to make code work correctly in the presence of mutation. JAX transforms still rely on functions being pure. I strongly suspect this is the root of your issue.

**Reply by samskiter** [reply]
Interesting. My code looks roughly like:

```[python]
class Thing:
    foo: float
    def get_error() -> float:
        return f(self.foo)
```

```[python]
class Optimizer:

    def __init__(self, things: [Thing]):
        self.things = things
        self.variables = [0.1] * len(things)

    def get_errors(things, variables):
        errors = jnp.empty(len(self.things)
        for idx, thing in enumerate(things):
            thing.foo = variables[idx] # <--- would this qualify as the side-effect ?
        for idx, thing in enumerate(things):
            errors = errors.at[idx].set[things.get_error()]
        return errors

    def jacobian(things, variables):
        return jaxfwd(Optimizer.optimise)(things, variables)

    def optimize_step(self):
        errors = Optimizer.get_errors(self.things, self.variables)
        jacobian = Optimizer.jacobian)(self.things, self.variables)
        inv_jacobian = jnp.pinv(jacobian)
        correction = jnp.matmul(inv_jacobian, errors)
        self.variables -= correction
        return errors.sum()

    def optimize(self)
        while self.optimize_step() < 1e-7
            print("optimize!")
```

I've tried to make the above 'pure'functions within the Optimizer class, but it is still storing variables on the `Thing` classes prior to calling a function on them.... I tried something like this setup in a minimal repro but didn't see the issue....

**Reply by jakevdp** [reply, marked as answer]
> `# <--- would this qualify as the side-effect ?`

Yes, definitely.

If you're updating attributes of `self` but not returning `self` from your function, then your function is impure and I would not expect JAX transformations to work with your code.

**Reply by samskiter** [reply]
@jakevdp yea, that's it - just nailed it down to a repro (below)! I wasn't expecting this to cause an issue after I had registered pytrees - my broad understanding was that allows my classes to be treated as if they were pure...

```
from __future__ import annotations
import jax
import jax.numpy as jnp
from jax.tree_util import register_pytree_node_class


@register_pytree_node_class  # <--- Even though Thing is registered as a pytree node, the jacobian function still fails
class Thing:
    foo: float

    def __init__(self, foo: float) -> None:
        self.foo = foo

    def tree_flatten(self) -> tuple[tuple[float], dict[str, float]]:
        return (self.foo,), {}

    @classmethod
    def tree_unflatten(
        cls, aux_data: dict[str, float], children: tuple[float]
    ) -> Thing:
        del aux_data
        return cls(children[0])

    def f(self, x: float) -> float:
        return x**2

    def my_f(self) -> float:
        return self.f(self.foo)

    def get_error(self) -> float:
        # return self.f(self.foo) # <--- this works
        return jax.lax.cond(  # <--- this fails
            True,
            self.my_f,
            self.my_f,
        )


class Optimizer:
    def __init__(self, things: [Thing]):
        self.things = things
        self.variables = jnp.array([0.1] * len(things))

    @staticmethod
    def get_errors(variables: jax.Array, things: list[Thing]) -> jax.Array:
        errors = jnp.empty(shape=(len(things)))
        for idx, thing in enumerate(things):
            thing.foo = variables[idx]  # <--- would this qualify as the side-effect ?

        print(f"things: {things}")
        for idx, thing in enumerate(things):
            error = thing.get_error()
            errors = errors.at[idx].set(error)
        return errors

    @staticmethod
    def jacobian(variables: jax.Array, things: [Thing]) -> jax.Array:
        return jax.jacfwd(Optimizer.get_errors, argnums=(0,))(variables, things)

    def optimize_step(self) -> float:
        print(f"variables: {self.variables}")
        print(f"things: {self.things}")
        errors = Optimizer.get_errors(self.variables, self.things)
        jacobian = Optimizer.jacobian(self.variables, self.things)
        print("jacobian", jacobian)
        exit(0)
        inv_jacobian = jnp.linalg.pinv(jacobian)
        correction = jnp.matmul(inv_jacobian, errors)
        self.variables -= correction
        return errors.sum()

    def optimize(self):
        while self.optimize_step() < 1e-7:
            print("optimize!")


optimizer = Optimizer([Thing(foo=0.1), Thing(foo=10.0)])
optimizer.optimize()

```

**Reply by samskiter** [reply]
Hrmm - yea re-reading the Jax documentation this is clear - often the docs are centered around @jit, but is then lumped in with grad and vmap, so I read the following sentence:

>This should work correctly with JIT and other transforms so long as you never mutate your object. Mutations of objects used as hash keys lead to several subtle problems, which is why for example mutable Python containers (e.g. [dict](https://docs.python.org/3/library/stdtypes.html#dict), [list](https://docs.python.org/3/library/stdtypes.html#list)) don’t define __hash__, while their immutable counterparts (e.g. [tuple](https://docs.python.org/3/library/stdtypes.html#tuple)) do.
<
>If your class relies on in-place mutations (such as setting `self.attr = ... within its methods), then your object is not really “static” and marking it as such may lead to problems. Fortunately, there’s another option for this case.
>
>Strategy 3: Making CustomClass a PyTree
The most flexible approach to correctly JIT-compiling a class method is to register the type as a custom PyTree object; see [Extending pytrees](https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees). This lets you specify exactly which components of the class should be treated as static and which should be treated as dynamic. Here’s how it might look:

and this: 

>Note that `grads` is a pytree with the same structure as `params`.
>`jax.grad` is one of the many JAX functions that has
> built-in support for pytrees.

(from different sections of the manual)

I got confused and thought pytrees would solve everything!

**Reply by samskiter** [reply]
https://github.com/google/jax/issues/1967 similarly

**Reply by jakevdp** [reply]
Yeah, I see where that could be confusing. But there's no getting around JAX's requirement that functions be pure.

**Reply by samskiter** [reply]
Damn, I spent like 2 days conforming my code to pytrees for this haha 😭

**Reply by jakevdp** [reply]
Pytree registration is not without value: for example you can now pass your objects to JIT-compiled functions, and you can JIT-compile your (pure) methods in-situ without having JIT cacheing bugs.

**Reply by samskiter** [reply]
@jakevdp thanks so much for your help - if you are open to it, I might have a think about a documentation change that might head off future users getting confused in the same way as me?

**Reply by jakevdp** [reply]
Sure, sounds good.

**Reply by samskiter** [reply]
Sorry to keep following up, I thought about it some more, and tried jitting the get_errors function and then the jacobian starts working!

If the below expected to work or am I just getting lucky?

It seems that by jitting the function, it is transformed into something that is more 'pure' ?

```
from __future__ import annotations
import jax
import jax.numpy as jnp
from jax.tree_util import register_pytree_node_class


@register_pytree_node_class  # <--- Even though Thing is registered as a pytree node, the jacobian function still fails
class Thing:
    foo: float

    def __init__(self, foo: float) -> None:
        self.foo = foo

    def tree_flatten(self) -> tuple[tuple[float], dict[str, float]]:
        return (self.foo,), {}

    @classmethod
    def tree_unflatten(
        cls, aux_data: dict[str, float], children: tuple[float]
    ) -> Thing:
        del aux_data
        return cls(children[0])

    def f(self, x: float) -> float:
        return x**2

    def my_f(self) -> float:
        return self.f(self.foo)

    def get_error(self) -> float:
        # return self.f(self.foo) # <--- this works
        return jax.lax.cond(  # <--- this fails
            True,
            self.my_f,
            self.my_f,
        )


class Optimizer:
    def __init__(self, things: [Thing]):
        self.things = things
        self.variables = jnp.array([0.1] * len(things))

    @staticmethod
    @jax.jit  # <--- Jitting this function makes it work
    def get_errors(variables: jax.Array, things: list[Thing]) -> jax.Array:
        errors = jnp.empty(shape=(len(things)))
        for idx, thing in enumerate(things):
            thing.foo = variables[idx]  # <--- would this qualify as the side-effect ?

        print(f"things: {things}")
        for idx, thing in enumerate(things):
            error = thing.get_error()
            errors = errors.at[idx].set(error)
        return errors

    @staticmethod
    def jacobian(variables: jax.Array, things: [Thing]) -> jax.Array:
        return jax.jacfwd(Optimizer.get_errors, argnums=(0,))(variables, things)

    def optimize_step(self) -> float:
        print(f"variables: {self.variables}")
        print(f"things: {self.things}")
        errors = Optimizer.get_errors(self.variables, self.things)
        jacobian = Optimizer.jacobian(self.variables, self.things)
        print("jacobian", jacobian)
        exit(0)
        inv_jacobian = jnp.linalg.pinv(jacobian)
        correction = jnp.matmul(inv_jacobian, errors)
        self.variables -= correction
        return errors.sum()

    def optimize(self):
        while self.optimize_step() < 1e-7:
            print("optimize!")


optimizer = Optimizer([Thing(foo=0.1), Thing(foo=10.0)])
optimizer.optimize()
```

**Reply by samskiter** [reply]
In short I'm saying `jacfwd(f)` gives different results to `jacfwd(jit(f, static_argnums(1,)))`. The latter works but the former fails.

**Reply by samskiter** [reply]
and, to further confirm the above seems to work - I tried it on my original code base and things seem to be working.

Line of argument:

- grad() requires pure functions
- a jit'd function is pure
- if a function is not pure, pytrees can be used to make it compatible with jit()
- if you jit a non-pure function, you end up with a pure function
- grad) can be used with pure functions if you first jit them

Am i wrong?

**Reply by jakevdp** [reply]
> if a function is not pure, pytrees can be used to make it compatible with jit()

This is not correct. Impure functions are not compatible with JIT or other JAX transformations.

> if you jit a non-pure function, you end up with a pure function

This is sort of correct: you do end up with a pure function, but it probably doesn't do the same thing as your original impure function (in particular, executing the JIT-compiled function will no longer result in the intended side-effect)

**Reply by samskiter** [reply]
>This is sort of correct: you do end up with a pure function, but it probably doesn't do the same thing as your original impure function (in particular, executing the JIT-compiled function will no longer result in the intended side-effect)

Good point. Interestingly I don't think I expected it to persist the side effect though.

IIUC what you're alluding to: the side effects / mutations wouldn't persist 'outside' of the jitted function, but would internally. In my example above, `get_errors` can mutate a `Thing` and then call `get_error` on that thing. But after `get_errors` returns, the `Thing` object wouldn't have the mutation applied. To be honest I don't think I expect it to though, because of how jitting/pytree flattening/unlfattening works - you can see the 'jit' as a barrier to mutation from within.

In my case I don't expect or require my mutations (which are just setting the attributes from the outer GD loop) to persists - I just want to use them to help me evaluate my gradient / error at a specific point in order to calculate my next GD step.

Am I understanding all of this correctly? It would be great if I could have `grad` / `jacfwd` have the same behaviour as jit here without having to use jit in between - that's where I really got tripped up!

**Reply by jakevdp** [reply]
To be honest I don't totally follow the details here, but I would suggest avoiding non-pure functions in JAX transformations. Even if you find cases where the semantics happen to be what you want, I can almost guarantee it will be the source of future bugs, because you're breaking the fundamental contract that JAX's logic is built on.

**Comment by samskiter** [comment]
Just an update - while what I'm doing may not be totally 'kosher' as per @jakevdp 's replies above, I think it's worth showing a little bit of the techniques I used to migrate an OO codebase to Jax. In the ends I created superclass to inherit and follow a pattern to allow my OO objects to be correctly flattened. This works with transient dependencies and is closely related to a previous proposal for pydags, here: https://github.com/google/jax/issues/7919 

Some notes on the approach:
1. I had to use the `register_pytree_node_class` approach on all my types. I used a common superclass to make this easier: Posted here: https://github.com/google/jax/issues/7919#issuecomment-1711475155
1. The above implement a flatten and unflatten function and place the 'uuid's of children into the auxilliary data.
1. using grad(f) didn't work but grad(jit(f)) did. Seems the JIT function is essential for forcing the flatten to be applied
1. I also used VMAP successfully, but I had to make sure I was doing vmap(f(x)) and *not* cmap(f(x,y)) - i.e. that there was only one argument - and that was a registered pytree that contained all the objects associated together
1. As we've mentioned in the other comments, mutations to objects will not persist outside of a jitted function - therefore you must return any values you want out of them. This suited me fine - I have objects with fixed properties and want to mutate some to produce an error function and perform gradient descent.

I'm actually overall pretty happy with what I've ended up with - while I wouldn't *start* from this place - adapting a code base to here has left me with something that's still quite neat without any really nasty hacks. The flatten/unflatten logic is still contained within each subclass and separates concerns nicely. Copilot is now able to write all my flatten/unflatten and getrefs/setrefs functions.

I would be interested in seeing something like this brought to Jax itself to make migration to Jax a little easier...
